keps/6757-failure-recovery/README.md

```diff
@@ -0,0 +1,341 @@
+as soon as possible, so I can make best use of the available resources (since the training process is synchronized,
+there is no risk of data corruption in my case).
+
+### Risks and Mitigations
+
+#### Replacing Pods Which Are Still Running
+
+Starting a replacement pod without a guarantee that the original one was properly terminated carries an inherent risk
+and might result in inconsistencies, as discussed above.
+If this risk is not communicated to the user properly and the feature is not sufficiently explicit in its behavior,
+the users might mistake it for a general issue within Kueue/the broader Kubernetes ecosystem.
+
+As this risk stems from the behavior of the `kubelet` and control plane themselves, it cannot be fully mitigated without major changes in how Kubernetes operates.
+Instead, it should be adequately documented to prevent users without a compatible use-case from using it.
+Moreover, enabling this feature should require opt-in both from:
+1. The Administrator - by defining the recovery rules and selectors in the Kueue config.
+2. The User - by opting specific workloads into the failure recovery mechanism.
+
+Controlling the covered workloads, instead of applying the recovery globally, will guarantee that only pods that were deemed "safe to forcefully terminate" are affected.
+
+## Design Details
+
+### Configuration API
+
+The `Configuration` struct is extended to add `FailureRecoveryPolicy`:
+
+```go
+type Configuration struct {
+  // ...
+
+  // FailureRecoveryPolicy is used to enable automatic failure recovery mechanisms.
+  // +optional
+  FailureRecoveryPolicy *FailureRecoveryPolicy `json:"failureRecoveryPolicy,omitempty"`
+}
+
+type FailureRecoveryPolicy struct {
+  // Rules specifies the rules to be enabled for failure recovery.
+  // +kubebuilder:validation:MinItems=1
+  Rules []FailureRecoveryRule `json:"rules"`
+}
+
+// +kubebuilder:validation:ExactlyOneOf=["terminatePod"]
+type FailureRecoveryRule struct {
+  // Exactly one of the fields below must be specified.
+
+  // TerminatePod enables and contains configuration for the `TerminatePod` strategy.
+  // This strategy recovers stuck pods by forcefully terminating them after a configured
+  // grace period elapses.
+  // +optional
+  TerminatePod *TerminatePodConfig `json:"terminatePod,omitempty"`
+}
+
+type TerminatePodConfig struct {
+  // PodLabelSelector specifies the scope of resources covered by `TerminatePod` failure recovery -
+  // resources not matching the selector are ignored by the controller.
+  // If a pod matches multiple configurations, the strictest one (with the shortest grace period)
+  // will be applied.
+  PodLabelSelector *metav1.LabelSelector `json:"podLabelSelector,omitempty"`
+
+  // ForcefulTerminationGracePeriod is the duration between when the pod's `deletionGracePeriodSeconds`
+  // elapses and when the pod should be forcefully deleted.
```

Comment:
This is more of the description about evaluating the order of rules, so the comment belongs to the "Rules" field inside FailureRecoveryPolicy.

However, I'm not certain this approach is the best. Alternatives I see:
1. we choose whatever is the first matching rule
2. we choose the least disruptive action first (say PodTermination over taining Node)
3. we choose the first in time action (as proposed currently)

I would suggest to avoid committing and for now only support a single rule.